contracts/contracts/strategies/ConvexOUSDMetaStrategy.sol

```diff
@@ -0,0 +1,223 @@
+// SPDX-License-Identifier: agpl-3.0
+pragma solidity ^0.8.0;
+
+/**
+ * @title Curve Convex Strategy
+ * @notice Investment strategy for investing stablecoins via Curve 3Pool
+ * @author Origin Protocol Inc
+ */
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+import { IRewardStaking } from "./IRewardStaking.sol";
+import { IConvexDeposits } from "./IConvexDeposits.sol";
+import { ICurvePool } from "./ICurvePool.sol";
+import { IERC20 } from "./BaseCurveStrategy.sol";
+import { BaseConvexMetaStrategy } from "./BaseConvexMetaStrategy.sol";
+import { StableMath } from "../utils/StableMath.sol";
+import { IVault } from "../interfaces/IVault.sol";
+
+contract ConvexOUSDMetaStrategy is BaseConvexMetaStrategy {
+    using StableMath for uint256;
+    using SafeERC20 for IERC20;
+
+    /* Take 3pool LP and mint the corresponding amount of ousd. Deposit and stake that to
+     * ousd Curve Metapool. Take the LP from metapool and deposit them to Convex.
+     */
+    function _lpDepositAll() internal override {
+        IERC20 threePoolLp = IERC20(pTokenAddress);
+        IERC20 metapoolErc20 = IERC20(address(metapool));
+        ICurvePool curvePool = ICurvePool(platformAddress);
+
+        uint256 threePoolLpBalance = threePoolLp.balanceOf(address(this));
+        uint256 curve3PoolVirtualPrice = curvePool.get_virtual_price();
+        uint256 threePoolLpDollarValue = threePoolLpBalance.mulTruncate(
+            curve3PoolVirtualPrice
+        );
+
+        uint128 crvCoinIndex = _getMetapoolCoinIndex(pTokenAddress);
+        uint128 ousdCoinIndex = _getMetapoolCoinIndex(
+            address(metapoolMainToken)
+        );
+
+        uint256 ousdToAdd = toPositive(
```

Comment:
When OUSD is low in the pool, this method can lose up to 0.04% of minted OUSD.

This is because curve is designed to have a large area where prices are near 1 to 1. As add catchup OUSD into this area, we end up not making a profit, but still paying a swap fee because our deposit is not balanced.

This can still happen on the other end of the situation as well. If the pool was balanced with a little too much 3pool, and we were putting more three pool tokens in, the ideal strategy probably would not be to put no OUSD in.

Reply:
Brownie console test run to show issue:

```python
from world import *
ousd_metapool = load_contract("ousd_metapool", OUSD_3CRVF_POOL)
threepool_virtual_price = threepool.get_virtual_price()

# NONFREN = '0x6c3F90f043a72FA612cbac8115EE7e52BDe6E490'
CONVEX = '0x25f0cE4E2F8dbA112D9b115710AC297F816087CD'


def pool_balances():
	print("%s 3pool, %s OUSD" % (
		c18(ousd_metapool.balances(1) * threepool_virtual_price // int(1e18)),
		c18(ousd_metapool.balances(0))
		)
	)

def amount_check():
	with TemporaryFork():
		ousd_metapool.remove_liquidity(ousd_metapool.balanceOf(flipper),[0,0], {'from': flipper})
		return history[-1].events['Transfer'][1]['_value'] * threepool_virtual_price // int(1e18) + history[-1].events['Transfer'][0]['value']

ousd.approve(ousd_metapool, 1e70, {'from': flipper})

amount = amount_check()
for i in range(0,10):
	ousd_metapool.add_liquidity([100000*1e18,0], 0, {'from':flipper}) 	
	new = amount_check()
	print(c18((new - amount) * 1000))
	amount = new


ousd_metapool.remove_liquidity_one_coin(1*1e6*1e18, 0, 0, governor, {'from': CONVEX})
ousd_metapool.remove_liquidity_one_coin(1*1e6*1e18, 0, 0, governor, {'from': CONVEX})
ousd_metapool.remove_liquidity_one_coin(1*1e6*1e18, 0, 0, governor, {'from': CONVEX})
ousd_metapool.remove_liquidity_one_coin(1*1e6*1e18, 0, 0, governor, {'from': CONVEX})



amount = amount_check()
for i in range(0,100):
	ousd_metapool.add_liquidity([10000*1e18,0], 0, {'from':flipper}) 	
	new = amount_check()
	print(c18(new - amount))
	amount = new
	pool_balances()
```

Reply:
I guess how this method works comes to which we prioritize more, peg holding, trading gain/loss, or long term APY. The current method is pure peg holding. Makes money if things are wildly out of peg, looses money if near peg, and does not prioritize APY/liquidity on one side (while it does on the other)

Reply:
Some time ago we have been discussing that we would want to maintain the metapool as balanced as possible. As in to provide stability to the pool. If we are to optimise for APY it would be better to print more OUSD to result in maximum amount of LP tokens. Similar things have been figured out in [this research](https://docs.google.com/spreadsheets/d/1DYSyYwHqxRzSJh9dYkY5kcgP_K5gku6N2mQVhoH33vY/edit#gid=0).

One way to modify the strategy to optimise APY (and consequently amount of LP deployed to Convex) would be to:
- always print at least the same amount of OUSD as counterpart 3crv
- if metapool low on OUSD (higher in 3pool) print so much OUSD that it is 50/50